File: day15/src/main/python/solution.py
import heapq


def read_input():
    f = open('../resources/input', 'r')
    y = -1
    risk_map = {}
    for line in f.readlines():
        y += 1
        x = -1
        for c in line.strip():
            x += 1
            risk_map[(x, y)] = int(c)
    return risk_map


def adjacent_coordinates(x, y):
    return [(x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)]


def find_max(maze):
    target = (0, 0)
    for key in maze:
        if key[0] > target[0] or key[1] > target[1]:
            target = key
    return target


def dijkstra(risk_map):
    queue = []
    heapq.heappush(queue, (0, (0, 0)))
    seen = set()
    target = find_max(risk_map)
    while len(queue) != 0:
        risk, current = heapq.heappop(queue)
        if current == target:
            return risk
        for option in adjacent_coordinates(current[0], current[1]):
            if option in risk_map and option not in seen:
                seen.add(option)
                heapq.heappush(queue, (risk + risk_map[option], option))


def grow(original):
    risk_map = {}
    x_range, y_range = find_max(original)
    x_range += 1
    y_range += 1
    for y_factor in range(5):
        for x_factor in range(5):
            for coordinate, risk in original.items():
                new_risk = risk + x_factor + y_factor
                risk_map[(coordinate[0] + x_factor * x_range, coordinate[1] + y_factor * y_range)] = \
                    new_risk if new_risk < 10 else new_risk - 10 + 1
    return risk_map


if __name__ == "__main__":
    risks = read_input()
    print("part 1")
    print(dijkstra(risks))
    print("part 2")
    print(dijkstra(grow(risks)))

File: day15/src/main/cpp/solution.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <string>
#include <vector>

using namespace std;
#define coordinate pair<int, int>
#define pic pair<int, coordinate>

map<coordinate, int> readInputFile() {
  ifstream file("../resources/input");
  map<coordinate, int> riskMap;
  string s;
  int y = -1;
  while (getline(file, s)) {
    ++y;
    int x = -1;
    for (auto const &risk : s) {
      riskMap[coordinate(++x, y)] = risk - '0';
    }
  }
  return riskMap;
}

vector<coordinate> adjacentCoordinates(coordinate c) {
  vector<coordinate> result;
  result.push_back(coordinate(c.first + 1, c.second));
  result.push_back(coordinate(c.first - 1, c.second));
  result.push_back(coordinate(c.first, c.second + 1));
  result.push_back(coordinate(c.first, c.second - 1));
  return result;
}

coordinate findMax(map<coordinate, int> &riskMap) {
  auto target = coordinate(0, 0);
  for (auto const &p : riskMap) {
    if (p.first.first > target.first || p.first.second > target.second) {
      target = p.first;
    }
  }
  return target;
}

int dijkstra(map<coordinate, int> riskMap) {
  set<coordinate> seen;
  seen.insert(coordinate(0, 0));
  priority_queue<pic, vector<pic>, greater<pic>> pqueue;
  pqueue.push(pair(0, coordinate(0, 0)));
  auto target = findMax(riskMap);
  while (!pqueue.empty()) {
    auto current = pqueue.top();
    pqueue.pop();
    if (current.second == target) {
      return current.first;
    }
    for (auto const &target : adjacentCoordinates(current.second)) {
      if (seen.insert(target).second && riskMap.find(target) != riskMap.end()) {
        pqueue.push(pair(current.first + riskMap[target], target));
      }
    }
  }
  throw 42;
}

map<coordinate, int> growMaze(map<coordinate, int> &riskMap) {
  auto maxRange = findMax(riskMap);
  int xRange = maxRange.first + 1;
  int yRange = maxRange.second + 1;
  coordinate c;
  map<coordinate, int> resultMap;
  for (int yFactor = 0; yFactor < 5; ++yFactor) {
    for (int xFactor = 0; xFactor < 5; ++xFactor) {
      for (auto const &p : riskMap) {
        auto risk = p.second + xFactor + yFactor;
        auto x = p.first.first + xFactor * xRange;
        auto y = p.first.second + yFactor * yRange;
        resultMap[coordinate(x, y)] = (risk < 10 ? risk : risk - 9);
      }
    }
  }
  return resultMap;
}

int main() {
  auto input = readInputFile();
  cout << "part 1" << endl
       << dijkstra(input) << endl
       << "part 2" << endl
       << dijkstra(growMaze(input)) << endl;
  return 0;
}
